File: IT/1/Funzioni/Esercizi/Java/Esercizio30.java
/* file Esercizio30.java */

import java.io.Console;
import java.util.function.IntBinaryOperator;
import java.util.function.BiFunction;
import java.util.Locale;

public class Esercizio30 {
	public static void main(String [] args) {
		Console console = System.console();
		// somma

		IntBinaryOperator add = (n1, n2) -> n1 + n2;

		// sottrazione
		IntBinaryOperator subtr = (n1, n2) -> n1 - n2;

		// prodotto
		IntBinaryOperator mult = (n1, n2) -> n1 * n2;

		// divisione
		BiFunction<Integer, Integer, Float> divide = (n1, n2) -> n2 != 0 ? (float) n1 / n2 : Float.NaN;

		String o1 = "?";
		String o2 = o1;

		while (!o1.matches("[+-]?[\\d]+")) {
			if (o1.isEmpty()) System.out.println("Stringa vuota!");
			else if (!o1.equals("?")) System.out.println("Numero \"" + o1 + "\" non valido!");
			o1 = console.readLine("Primo operando: ");
		}

		while (!o2.matches("[+-]?[\\d]+")) {
			if (o2.isEmpty()) System.out.println("Stringa vuota!");
			else if (!o2.equals("?")) System.out.println("Numero \"" + o2 + "\" non valido!");
			o2 = console.readLine("Secondo operando: ");
		}

		int n1 = Integer.parseInt(o1);
		int n2 = Integer.parseInt(o2);

		String op = System.console().readLine("Scegli un'operazione:\n+ - somma,\n- - sottrazione,\n* - prodotto,\n/ - divisione\n");

		if (op.equals("+")) {
			System.out.println(n1 + " + " + n2 + " = " + add.applyAsInt(n1, n2));
		}

		if (op.equals("-")) {
			System.out.println(n1 + " - " + n2 + " = " + subtr.applyAsInt(n1, n2));
		}

		if (op.equals("*")) {
			System.out.println(n1 + " * " + n2 + " = " + mult.applyAsInt(n1, n2));
		}

		if (op.equals("/")) {
			System.out.println(n1 + " / " + n2 + " = " + String.format(Locale.ROOT, "%.3f", divide.apply(n1, n2)));
		}
	}
}

File: IT/1/Funzioni/Esercizi/Python/esercizio30.py
# file esercizio30.py

from __future__ import print_function, division
import re

if hasattr(__builtins__, 'raw_input'): input = raw_input

add = lambda n1, n2 : n1 + n2

subtr = lambda n1, n2 : n1 - n2

mult = lambda n1, n2: n1 * n2

divide = lambda n1, n2: n1 / n2 if n2 != 0 else float('nan')

o1 = "?"
o2 = o1

while not re.match("[+-]?[\d]+", o1):
	if not o1: print("Stringa vuota!")
	elif o1 != "?": print("Numero \"{}\" non valido!".format(o1))
	o1 = input("Primo operando: ")

while not re.match("[+-]?[\d]+", o2):
	if not o2: print("Stringa vuota!")
	elif o2 != "?": print("Numero \"{}\" non valido!".format(o2))
	o2 = input("Secondo operando: ")

n1 = int(o1)
n2 = int(o2)

op = input("Scegli un'operazione:\n+ - somma,\n- - sottrazione,\n* - prodotto,\n/ - divisione\n")

result = 0

if op == "+":
	print("{} + {} = {}".format(n1, n2, add(n1, n2)));

if op == "-":
	print("{} - {} = {}".format(n1, n2, subtr(n1, n2)));

if op == "*":
	print("{} * {} = {}".format(n1, n2, mult(n1, n2)))

if op == "/":
	print("{} / {} = {:.3f}".format(n1, n2, divide(n1, n2)))

